reduction/reduction.cu

```diff
@@ -0,0 +1,152 @@
+// Engineer: Abhishek Gautam
+// Last Updated: 02-22-2025
+
+// Problem/Functionality: This is a Kernel to perform a reduction on an input
+//                        array. In this case we are simply going to sum all
+//                        elements of the array.
+
+// High-level Approach:
+// 1) The array is to be divided into segments processed by thread blocks of
+//    a particular maximum size.
+// 2) A tree based approach is used within each thread block.
+//     2.1) Each block has as many elements to work with as threads.
+//     2.2) A loop is run in which in each cycle the threads reduce the
+//          block data to half its size until only 1 element is left which
+//          is the result of the reduction on the block
+//     2.3) The block's result is stored at its block index in the input
+//          array.
+// 3) The problem size has now shrunk from N to no. of blocks launced. This
+//    process along with the kernel launch is repeated in a loop on the
+//    results of the thread blocks until 1 thread block is left which
+//    eveluates to one result which is the final result.
+// 4) As an optimization, each thread brings in two values from the input
+//    array into shared memory. Thus, each block processes twice the size
+//    of input elements. This helps distribute work among the threads more
+//    evenly threads start to become idle as the process continues.
+// 5) As an optimization, the last warp(32 threads) worth of work is
+//    unrolled. This helps avoid instruction overhead and reduces useless
+//    work in the other warps.
+
+// =============================================================================
+
+#include <stdio.h>
```

Comment:
I assume you can use C++17 or newer here, in which case you may want to modernize it to avoid C-isms
Here: include <cstdio>